File: tests/UniformMultiLayerLeverDrive2d.py
import unittest

import FrictionQPotFEM
import GMatElastoPlasticQPot.Cartesian2d as GMat
import GooseFEM
import numpy as np


class test_Generic2d(unittest.TestCase):
    """
    Tests
    """

    def test_eventDrivenSimpleShear(self):

        mesh = GooseFEM.Mesh.Quad4.Regular(nx=3, ny=2 * 3 + 1)
        nelem = mesh.nelem()
        coor = mesh.coor()
        conn = mesh.conn()
        elem = mesh.elementgrid()
        dofs = mesh.dofs()
        dofs[mesh.nodesLeftEdge()[1:], ...] = dofs[mesh.nodesRightEdge()[1:], ...]
        layers = [elem[:3, :].ravel(), elem[3, :].ravel(), elem[4:, :].ravel()]
        layers = [np.sort(i) for i in layers]

        system = FrictionQPotFEM.UniformMultiLayerLeverDrive2d.System(
            coor,
            conn,
            dofs,
            dofs[mesh.nodesBottomEdge(), :].ravel(),
            layers,
            [np.unique(conn[i, :]) for i in layers],
            [False, True, False],
        )

        nelas = system.elastic().size
        nplas = system.plastic().size

        Hi = []
        for layer in layers:
            yl = coor[conn[layer[0], 0], 1]
            yu = coor[conn[layer[-1], 3], 1]
            Hi += [0.5 * (yu + yl)]
        H = np.max(coor[:, 1])

        epsy = 1e-3 * np.cumsum(np.ones((nplas, 5)), axis=1)

        system.setMassMatrix(np.ones(nelem))
        system.setDampingMatrix(np.ones(nelem))
        system.setElastic(np.ones(nelas), np.ones(nelas))
        system.setPlastic(np.ones(nplas), np.ones(nplas), epsy)
        system.setDt(1.0)
        system.layerSetDriveStiffness(1e-3)
        system.setLeverProperties(H, Hi)

        drive_active = np.zeros((3, 2), dtype=bool)
        drive_active[-1, 0] = True

        for loop in range(2):

            if loop == 0:
                system.initEventDriven(0.1, drive_active)
                delta_xlever = system.eventDriven_deltaLeverPosition()
                delta_active = system.eventDriven_targetActive()
                delta_u = system.eventDriven_deltaU()
                delta_ubar = system.eventDriven_deltaUbar()
            else:
                system.initEventDriven(delta_xlever, drive_active, delta_u, delta_ubar)
                system.setU(np.zeros_like(coor))

            system.eventDrivenStep(1e-4, False)
            self.assertTrue(np.allclose(GMat.Epsd(system.plastic_Eps()), epsy[0, 0] - 0.5 * 1e-4))
            self.assertTrue(system.residual() < 1e-5)

            system.eventDrivenStep(1e-4, False)
            self.assertTrue(np.allclose(GMat.Epsd(system.plastic_Eps()), epsy[0, 0] - 0.5 * 1e-4))
            self.assertTrue(system.residual() < 1e-5)

            system.eventDrivenStep(1e-4, True)
            self.assertTrue(np.allclose(GMat.Epsd(system.plastic_Eps()), epsy[0, 0] + 0.5 * 1e-4))
            self.assertTrue(system.residual() < 1e-5)

            system.eventDrivenStep(1e-4, False)
            self.assertTrue(np.allclose(GMat.Epsd(system.plastic_Eps()), epsy[0, 1] - 0.5 * 1e-4))
            self.assertTrue(system.residual() < 1e-5)

            system.eventDrivenStep(1e-4, True)
            self.assertTrue(np.allclose(GMat.Epsd(system.plastic_Eps()), epsy[0, 1] + 0.5 * 1e-4))
            self.assertTrue(system.residual() < 1e-5)

            system.eventDrivenStep(1e-4, False)
            self.assertTrue(np.allclose(GMat.Epsd(system.plastic_Eps()), epsy[0, 2] - 0.5 * 1e-4))
            self.assertTrue(system.residual() < 1e-5)

            system.eventDrivenStep(1e-4, False, -1)
            self.assertTrue(np.allclose(GMat.Epsd(system.plastic_Eps()), epsy[0, 1] + 0.5 * 1e-4))
            self.assertTrue(system.residual() < 1e-5)

            system.eventDrivenStep(1e-4, True, -1)
            self.assertTrue(np.allclose(GMat.Epsd(system.plastic_Eps()), epsy[0, 1] - 0.5 * 1e-4))
            self.assertTrue(system.residual() < 1e-5)

            system.eventDrivenStep(1e-4, False, -1)
            self.assertTrue(np.allclose(GMat.Epsd(system.plastic_Eps()), epsy[0, 0] + 0.5 * 1e-4))
            self.assertTrue(system.residual() < 1e-5)


if __name__ == "__main__":

    unittest.main()

File: tests/UniformMultiLayerLeverDrive2d.cpp
#include <FrictionQPotFEM/UniformMultiLayerLeverDrive2d.h>
#include <GooseFEM/MeshQuad4.h>
#include <catch2/catch_all.hpp>
#include <xtensor/xrandom.hpp>

TEST_CASE("FrictionQPotFEM::UniformMultiLayerLeverDrive2d", "UniformMultiLayerLeverDrive2d.h")
{
    SECTION("Force from drive: prescribe mean of each layer")
    {
        GooseFEM::Mesh::Quad4::Regular mesh(5, 1);

        xt::xtensor<bool, 1> is_plastic = {false, true, false, true, false};
        size_t nlayer = is_plastic.size();
        GooseFEM::Mesh::Vstack stitch;
        for (size_t i = 0; i < nlayer; ++i) {
            stitch.push_back(mesh.coor(), mesh.conn(), mesh.nodesBottomEdge(), mesh.nodesTopEdge());
        }

        auto dofs = stitch.dofs();
        auto bottom = stitch.nodeset(mesh.nodesBottomEdge(), 0);
        auto iip = xt::ravel(xt::view(dofs, xt::keep(bottom)));
        auto layers = stitch.elemmap();

        size_t nelas = 0;
        size_t nplas = 0;

        for (size_t i = 0; i < layers.size(); ++i) {
            if (is_plastic[i]) {
                nplas += layers[i].size();
            }
            else {
                nelas += layers[i].size();
            }
        }

        xt::xtensor<bool, 2> drive = xt::zeros<bool>({5, 2});
        xt::xtensor<double, 2> u_target = xt::zeros<double>({5, 2});
        xt::xtensor<double, 1> height = 0.5 * xt::arange<double>(5);

        drive(2, 0) = true;
        drive(4, 0) = true;

        u_target(2, 0) = height(2);
        u_target(4, 0) = height(4);

        FrictionQPotFEM::UniformMultiLayerLeverDrive2d::System sys(
            stitch.coor(),
            stitch.conn(),
            dofs,
            iip,
            layers,
            stitch.nodemap(),
            is_plastic,
            FrictionQPotFEM::moduli_toquad(xt::eval(xt::ones<double>({nelas}))),
            FrictionQPotFEM::moduli_toquad(xt::eval(xt::ones<double>({nelas}))),
            FrictionQPotFEM::moduli_toquad(xt::eval(xt::ones<double>({nplas}))),
            FrictionQPotFEM::moduli_toquad(xt::eval(xt::ones<double>({nplas}))),
            FrictionQPotFEM::epsy_initelastic_toquad(100 * xt::ones<double>({nplas, size_t(1)})),
            1,
            1,
            1,
            0,
            drive,
            1.0,
            height.back(),
            height);

        sys.layerSetUbar(u_target, drive);
        sys.setLeverTarget(1.0 * height.back());

        REQUIRE(sys.leverPosition() == Catch::Approx(sys.leverTarget()));
        REQUIRE(xt::allclose(sys.fdrive(), 0.0));
        REQUIRE(xt::allclose(sys.layerFdrive(), 0.0));
    }
}
